numpy/core/tests/test_multiarray.py

```diff
@@ -451,6 +451,14 @@ def test_longdouble_assignment(self):
             arr = np.array([np.array(tinya)])
             assert_equal(arr[0], tinya)
 
+    def test_cast_to_string(self):
+        # cast to str should do "str(scalar)", not "str(scalar.item())"
+        # Example: In python2, str(float) is truncated, so we want to avoid
+        # str(np.float64(...).item()) as this would incorrectly truncate.
+        a = np.zeros(1, dtype='S20')
+        a[:] = np.array(['1.12345678901234567890'], dtype='f8')
+        assert_equal(a[0], b"1.1234567890123457")
```

Comment:
Can you add another test for `S10` too?

It would be good to record what happens when the float str is too long. I'm not sure whether letting the string truncate is OK, or if it should round.

Reply:
Also might be tidier to use a 0d array here and `.astype`, but up to you

Reply:
For all types (not just floats) there is a fairly well established precedent that strings get simply truncated if the dtype is too short, so I don't think we should round. Integers don't round.

Also, I'm a bit hesitant to add a test for `S10` because I didn't touch any string truncation code here, and getting the truncation right seems like a separate issue, to be examined and discussed on its own. This PR's purpose is to test string casting intermediates, not to test the arguably unrelated string truncation.